### ISSUE TITLE
lint: do not compare enum types

### ISSUE THREAD

**Issue by cristaloleg** [go-critic/go-critic#442]
Name: `enumCompare`
Before:
```go
type myType int
var a myType = myValue1
if a >= myValue1 && a < myValue3 { ... }
```
After:
```go
type myType int
var a myType = myValue1
if a == myValue && a == myValue2 { ... }
```

#VeryOpinionated

**Comment by Quasilyte** [comment]
Yep, very opinionated.
I think it's OK to compare for range check.
For example:
```go
if x >= firstStmtKind && x <= lastStmtKind {}
```
Very common pattern.
Maybe you should avoid warning if there is conventional `begin`+`end` name parts and/or `first`/`last`.
Could reduce false positives.

**Comment by Quasilyte** [comment]
Can you provide real world code example, where this "enum" comparison makes things worse?

**Comment by cristaloleg** [comment]
No, nothing from real world. But hiding values between comparison operators make code less clear. At least for me 😉

**Comment by Quasilyte** [comment]
Hmm. I don't get how it does value hidding.
It's category check.
Used a whole lot inside golang/go code.

**Comment by Quasilyte** [comment]
There can be dozens expression kinds, it would be awkward (and fragile) to check for every one of them directly.

**Comment by cristaloleg** [comment]
```go
type myType int

const (
    valueFoo myType = iota
    valueBar
    valueBar10
    valueBar20
    valueBar30
    valueBaz
)

a >= valueFoo && a < valueBaz  // <- not so clear what is in between
```

**Comment by Quasilyte** [comment]
Well.. OK, but this will have 99% of false positives that should not be fixed.
You should checkout code that uses this pattern.
It's quite weird to require rewriting into more fragile and more error-prone form (one can do a typo or something).

**Comment by Quasilyte** [comment]
We can include it as opinionated, but the problem is that it checks for valid code that is kind of idiomatic.

**Comment by Quasilyte** [comment]
For baz+foo constants maybe there is a mistake, if those values in range are unrelated.
